Elaborate function application sorts in `rty`

ranjitjhala opened the pull request:
1. Changes the `rty::Expr::App` to explicitly track the `List<SortArg>` that were inferred during the `sortck`
2. Generalizes `char_to_int` to "polymorphic" ~~`to_int`~~ `cast`
3. Uses the explicit `SortArg` during fixpoint-conversion for ~~`to_int`~~ `cast`.

Allows stuff like

```rust
use flux_rs::attrs::*;

#[spec(fn (x:u8) -> u32[cast(x)])]
pub fn test_int_to_int(x: u8) -> u32 {
    x as u32
}

#[spec(fn (x:char) -> u32[cast(x)])]
pub fn test_char_to_int(x: char) -> u32 {
    x as u32
}

#[spec(fn (x:bool) -> u32[cast(x)])]
pub fn test_bool_to_int(x: bool) -> u32 {
    x as u32
}

#[spec(fn (x:bool) -> u32[cast(x)])]
pub fn test_bool_to_int_with_if(x: bool) -> u32 {
    if x { 1 } else { 0 }
}

pub struct S {}

#[spec(fn(x:S, n:usize[to_int(x)]) -> usize[cast(x)])]
pub fn foo(_x: S, n: usize) -> usize {
    n
}
```

nilehmann commented on crates/flux-infer/src/fixpoint_encoding.rs:
```diff
@@ -1459,6 +1484,30 @@ impl<'genv, 'tcx> ExprEncodingCtxt<'genv, 'tcx> {
         }
     }
 
+    fn define_const_for_to_int(
```
```suggestion
    fn define_const_for_cast(
```

nilehmann commented on crates/flux-infer/src/fixpoint_encoding.rs:
```diff
@@ -377,6 +377,7 @@ enum ConstKey<'tcx> {
     Alias(FluxDefId, rustc_middle::ty::GenericArgsRef<'tcx>),
     Lambda(Lambda),
     PrimOp(rty::BinOp),
+    ToInt(List<rty::SortArg>),
```
```suggestion
    Cast(List<rty::SortArg>),
```

we could also make this `Cast(rty::Sort, rty::Sort)` to be very explicit.

ranjitjhala replied:
Ok got it. I think its fine to generate generic UIFs in fixpoint too... I don't think this matters too much one way or another. (The bigger issue is whether to use UIFs _at all_ in this case, or just disallow...)

nilehmann replied:
I was thinking that if we stop emitting generic UIFs that would make the transition to the fixpoint Rust reimplementation easier, but that's a long way ahead...

nilehmann commented on crates/flux-middle/src/fhir.rs:
```diff
@@ -1193,7 +1193,7 @@ pub enum SpecFuncKind {
     Def(FluxDefId),
     /// Char-Int Conversions
-    CharToInt,
+    Cast,
```
Adjust documentation

nilehmann commented on crates/flux-middle/src/sort_of.rs:
```diff
@@ -41,6 +43,22 @@ impl GlobalEnv<'_, '_> {
         if ty.is_integral() { Ok(Some(rty::Sort::Int)) } else { self.sort_of_rust_ty(def_id, ty) }
     }
 
+    pub fn sort_of_spec_func(self, spec_func: &SpecFuncKind) -> rty::PolyFuncSort {
+        match spec_func {
+            SpecFuncKind::Def(name) | SpecFuncKind::Uif(name) => self.func_sort(name),
+            SpecFuncKind::Thy(itf) => THEORY_FUNCS.get(itf).unwrap().sort.clone(),
+            SpecFuncKind::Cast => {
+                rty::PolyFuncSort::new(
+                    List::from_arr([SortParamKind::Sort, SortParamKind::Sort]),
+                    rty::FuncSort::new(
+                        vec![rty::Sort::Var(rty::ParamSort::from(0_usize))],
+                        rty::Sort::Var(rty::ParamSort::from(1_usize)),
+                    ),
+                )
+            }
+        }
+    }
+
```
Please revert this change. At some point, I want to stop treating theory functions as "functions" and have dedicated rules for them, so we can get rid of parametric bit vector sizes. This change goes in the wrong direction.

nilehmann commented on crates/flux-infer/src/fixpoint_encoding.rs:
```diff
@@ -1459,6 +1484,30 @@ impl<'genv, 'tcx> ExprEncodingCtxt<'genv, 'tcx> {
+        self.const_map
+            .entry(key)
+            .or_insert_with(|| {
+                let fsort = self
+                    .genv
+                    .sort_of_spec_func(&flux_middle::fhir::SpecFuncKind::Cast)
+                    .instantiate(sort_args);
+                let fsort = rty::PolyFuncSort::new(List::empty(), fsort);
+                let sort = scx.func_sort_to_fixpoint(&fsort);
```
Since I suggested getting rid of `sort_of_spec_func`, we can inline the function sort here. We are already relying on `cast` having exactly one argument for the encoding, so this is already a leaky abstraction, and we don't lose much with the extra "coordination".

nilehmann commented on book/src/guide/specifications.md:
````diff
@@ -957,3 +957,11 @@ or lemmas. See `tests/tests/pos/surface/primops00.rs` for an example.
 {{#include ../../../tests/tests/pos/surface/primops00.rs}}
 ```
+
+## Casting Sorts to Int
+
+You can convert refinements of different sorts to `int` using the `to_int` internal function.
+
+```rust,noplayground
+{{#include ../../../tests/tests/pos/surface/to_int00.rs}}
+```
````
Update documentation to use `cast`

nilehmann commented on tests/tests/pos/surface/to_int00.rs:
```diff
@@ -1,12 +1,16 @@
-        let i = char_to_int(c);
-        48 <= i && i <= 57
+        let i = cast(c);
+        as_int(48) <= i && i <= as_int(57)
```
Hmm, I thought we could use an `int` postfix to force an integer, but even though we are parsing it correctly, it is not working as intended.

```
        48int <= i && i <= 57int
```
We could also implement a fallback behavior for unconstrained num vars so they become int.

ranjitjhala replied:
should I try to implement this *fallback* in this PR or separately?

ranjitjhala replied:
Hmm its a bit more complicated as we have to also treat `48 <= i` as an `int` operation.

Why is `48` given a `NumVar` in the first place (instead of just being `int` ?

nilehmann replied:
Int literals can be treated as bit vectors, ints, or reals dependening on they inferred use.

The way the fallback would work is that we record the places where we generate a NumVar for an int literal. Then, at the end of sort checking we check if they are solved and if they are not we unify them with int. That should fix the sort of binary relations as well.

ranjitjhala replied:
Right now it seems that we only generate `NumVar` for int-literals. When you do what you're suggesting -- the `fully_resolve` check inside `into_results` ? You can turn the literals to `int` but the `sort_of_bin_op` remains unresolved...? So somehow you'd have to "re-propagate" the int-ness there too, no?

nilehmann commented on crates/flux-infer/src/fixpoint_encoding.rs:
```diff
@@ -1064,10 +1072,27 @@ impl<'genv, 'tcx> ExprEncodingCtxt<'genv, 'tcx> {
+                        );
+                        self.expr_to_fixpoint(&expr, scx)
+                    }
+                    // Treat all other to_int conversions as uninterpreted
```
I'm not a fan of casts being silently "ignored" as uninterpreted functions. That's a recipe for "why doesn't my code verify?"

But on the plus side, if everything is uniformly castable, then you can have, for example, a function generic on a sort that is castable to `int` without having explicit bounds.

Maybe we can have the cake and eat it too, if we reject statically casts that don't make sense when both sides are concrete.

nilehmann replied:
Yes, they are genuine use cases for the uif. The problem is when it's unintended.

This could be implemented as a warn-by-default lint that can be silenced. But we don't have any other examples of that so I don't know how hard that would be.

ranjitjhala replied:
How about we have error-by-default but you instead can have a flag to `allow_uninterpreted_cast` ?

nilehmann replied:
Yes, that would have the intended behavior. We could even make it per item if we add it as an option to `InferOpts`.

nilehmann replied:
Also, there are two ways in which one could fail. 

1. During sort checking
2. During fixpoint encoding

I'm thinking about 1.

ranjitjhala commented:
@nilehmann I addressed the comments, but there are two remainders:

1. Remove the UIF for non-supported casts.
2. Implement the default-int-literal.

Thoughts?

ranjitjhala commented:
@nilehmann lmk if there is more to be done here..
